### PR TITLE
Fix Java incremental compilation corner cases related to generics

### LINKED ISSUE #20478: Incremental java compiler does not detect changes to classes used in type parameters in some cases
### Expected Behavior
Gradle build should fail when Java compilation should fail.

### Current Behavior
Gradle build succeeds when Java compilation should fail.

### Context

Given these classes:
```
interface Super <T> { 
	T getResult();
}

interface Sub extends Super<Deleted> { }

interface Deleted {}
```
If I delete `Deleted`, `Sub` should be recompiled (and fail compilation). I found this in the gradle/gradle codebase when deleting some classes.

The build fails if `Sub` has a reference to `T` or I disable incremental compilation.

A similar thing happens in a slightly simpler example:
```
interface Sub<T extends Deleted> { }

interface Deleted {}
```
It looks like we ignore the type parameters in some cases.

### Steps to Reproduce 

I reproduced the failure in a test:
https://github.com/gradle/gradle/pull/20479

### Your Environment

This happens everywhere.

I tested 5.0, 6.0, 7.0 and 7.3/7.4/7.5. I think this has never worked.

### PR DESCRIPTION
**Pull request by jbartok**
Fixes #20478 

Doesn't solve however the same problem if the generic type belongs to a static method, not the class. Have added a test for this scenario too.

### CODE REVIEW

**Comment by jbartok** [comment]
@bot-gradle test QFL

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/asm/ClassDependenciesVisitor.java [hunk 1]
```diff
@@ -73,6 +75,16 @@ public static ClassAnalysis analyze(String className, ClassReader reader, String
...
+        if (signature != null) {
+            SignatureReader signatureReader = new SignatureReader(signature);
+            signatureReader.accept(new SignatureVisitor(API) {
+                @Override
+                public void visitClassType(String className) {
+                    Type type = Type.getObjectType(className);
+                    maybeAddDependentType(types, type);
+                }
+            });
+        }
```
Comment (asodja) [inline subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/asm/ClassDependenciesVisitor.java:87]: Nitpick: could we move this part to some method, like `visitSignature()`?

File: subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/JavaClassChangeIncrementalCompilationIntegrationTest.groovy [hunk 2]
```diff
@@ -421,4 +422,22 @@ class JavaClassChangeIncrementalCompilationIntegrationTest extends BaseJavaClass
         deleted.delete()
         recompiledWithFailure('symbol: class Deleted', 'Sub')
     }
+
+    @NotYetImplemented
```
Comment (asodja) [inline subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/JavaClassChangeIncrementalCompilationIntegrationTest.groovy:426]: I also wonder what happens if we have a generic on a field of a class, e.g.:
```
class A {
    List<Deleted> list;
}
```

File: subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/JavaClassChangeIncrementalCompilationIntegrationTest.groovy [hunk 3]
```diff
@@ -409,9 +409,10 @@ class JavaClassChangeIncrementalCompilationIntegrationTest extends BaseJavaClass
         outputs.recompiledClasses('A', 'B')
     }
 
-    @NotYetImplemented
     @Issue("https://github.com/gradle/gradle/issues/20478")
-    def "can detect deletion of class used as type parameter"() {
+    def "can detect deletion of class used as class type parameter"() {
+        //executer.requireOwnGradleUserHomeDir()  //use when debugging to bypass cache
+
         source "interface Super <T> { T getResult(); }"
```
Comment (asodja) [inline subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/JavaClassChangeIncrementalCompilationIntegrationTest.groovy:416]: Could we also add test just to be sure, when class have multiple parameters, e.g. `Super<Deleted, Deleted2>`?

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/asm/ClassDependenciesVisitor.java [hunk 4]
```diff
@@ -73,6 +75,16 @@ public static ClassAnalysis analyze(String className, ClassReader reader, String
     public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
         isAnnotationType = isAnnotationType(interfaces);
         Set<String> types = isAccessible(access) ? accessibleTypes : privateTypes;
+        if (signature != null) {
+            SignatureReader signatureReader = new SignatureReader(signature);
+            signatureReader.accept(new SignatureVisitor(API) {
+                @Override
+                public void visitClassType(String className) {
+                    Type type = Type.getObjectType(className);
+                    maybeAddDependentType(types, type);
```
Comment (asodja) [inline subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/asm/ClassDependenciesVisitor.java:84]: I wonder if generics are part of ABI or we could always add them to `privateTypes`?

**Comment by jbartok** [comment]
@bot-gradle test RFM

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/asm/ClassDependenciesVisitor.java [hunk 5]
```diff
@@ -218,26 +235,27 @@ protected MethodVisitor(Set<String> types) {
...
         @Override
         public void visitLocalVariable(String name, String desc, String signature, Label start, Label end, int index) {
-            maybeAddDependentType(types, Type.getType(desc));
+            maybeAddClassTypesFromSignature(signature, privateTypes);
+            maybeAddDependentType(privateTypes, Type.getType(desc));
             super.visitLocalVariable(name, desc, signature, start, end, index);
         }
 
         @Override
         public org.objectweb.asm.AnnotationVisitor visitAnnotation(String descriptor, boolean visible) {
```
Comment (asodja) [inline subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/asm/ClassDependenciesVisitor.java:244]: Sorry for trouble with that but can we revert these? `visitLocalVariable ` is ok, but these `visitAnnotation`, `visitParameterAnnotation` and  `visitParameterAnnotation ` might not be.

I checked on some demo Micronaut project and annotation processor checks also annotations for parameters from super type. So I believe this change would not be ok for annotation processor that needs some more info from super class.

**Review by asodja — APPROVED** [review]
LGTM, just apply a change for the comment about  `visitAnnotation`, `visitParameterAnnotation` and `visitParameterAnnotation`

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/asm/ClassDependenciesVisitor.java (hunk 5 above)
Comment (jbartok) [reply subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/asm/ClassDependenciesVisitor.java:244]: Done.

**Comment by jbartok** [comment]
@bot-gradle test and merge

**Review by asodja — APPROVED** [review]

**Comment by jbartok** [comment]
@bot-gradle test and merge

**Comment by jbartok** [comment]
@bot-gradle test and merge

**Pull request merged**